### PR TITLE
Containerregistry cached

### PR DESCRIPTION
**Pull request by smukherj1**

### CODE REVIEW

**Comment by nlopezgi** [comment]
Hi @anchlovi,
Could you please confirm here in this PR that its OK to use your previous commits? (just a message here saying its OK will do) Thanks!

File: README.md [hunk 1]
```diff
@@ -1139,6 +1139,18 @@ configuration. See [here](#container_pull-custom-client-configuration) for detai
...
+**NOTE:** Set `DOCKER_REPO_CACHE` env variable to make the container puller
+cache downloaded layers at the directory specified as a value to this env
+variable. The caching feature hasn't been thoroughly tested and may be thread
+unsafe. If you notice flakiness after enabling it, see the warning below on how
+to workaround it.
+
+<aside class="warning">
+`container_pull` is suspected to have thread safety issues. To ensure multiple
+container_pull(s) don't execute concurrently, please use the bazel startup
+flat `--loading_phase_threads=1` in your bazel invocation.
```
Comment (nlopezgi) [inline README.md]: flag

File: container/pull.bzl [hunk 2]
```diff
@@ -185,4 +195,5 @@ pull = struct(
 container_pull = repository_rule(
     attrs = _container_pull_attrs,
     implementation = _impl,
+    environ = ["DOCKER_REPO_CACHE", "PULLER_TIMEOUT", "HOME"],
```
Comment (nlopezgi) [inline container/pull.bzl]: One per line in alphabetical order

File: testing/e2e.sh [hunk 3]
```diff
@@ -500,6 +500,28 @@ function test_container_push_all() {
...
+  rm -rf $scratch_dir
+  mkdir -p $cache_dir
+  mkdir -p $bazel_cache
+
+  # Run container puller one with caching.
+  DOCKER_REPO_CACHE=$cache_dir PULLER_TIMEOUT=600 bazel --output_base=$bazel_cache test //tests/docker:distoless_fixed_id_digest_test
+
+  # Rerun the puller by changing the puller timeout to force a rerun of of the
+  # target but now using the cache instead of downloading it again.
+  DOCKER_REPO_CACHE=$cache_dir PULLER_TIMEOUT=601 bazel --output_base=$bazel_cache test //tests/docker:distoless_fixed_id_digest_test
```
Comment (nlopezgi) [inline testing/e2e.sh]: how do we know it worked?

**Comment by anchlovi** [comment]
> Hi @anchlovi,
> Could you please confirm here in this PR that its OK to use your previous commits? (just a message here saying its OK will do) Thanks!

Sure

File: README.md (hunk 1 above)
Comment (smukherj1) [reply README.md]: Done

File: container/pull.bzl (hunk 2 above)
Comment (smukherj1) [reply container/pull.bzl]: Done

File: testing/e2e.sh (hunk 3 above)
Comment (smukherj1) [reply testing/e2e.sh]: As discussed offline there's no easy way to determine this. To ensure the cached pull is faster, we need to setup a full fledged benchmarking that runs several samples and takes an average

**Comment by smukherj1** [comment]
Setting the cla to yes manually as @anchlovi has ok'd the use of his commits in this PR

**Comment by nlopezgi** [comment]
Disabled CLA check to merge. original commit author gave OK in message above.

**Pull request merged**